File: Estructurales Java/Flyweight.java
// Flyweight Pattern
import java.util.HashMap;
import java.util.Map;
import java.util.List;
import java.util.Arrays;

interface Flyweight {
    String operation(String extrinsicState);
}

class ConcreteFlyweight implements Flyweight {
    private String intrinsicState;
    
    public ConcreteFlyweight(String intrinsicState) {
        this.intrinsicState = intrinsicState;
    }
    
    @Override
    public String operation(String extrinsicState) {
        return "ConcreteFlyweight: Intrinsic(" + intrinsicState + "), Extrinsic(" + extrinsicState + ")";
    }
}

class FlyweightFactory {
    private Map<String, Flyweight> flyweights = new HashMap<>();
    
    public FlyweightFactory(List<String> initialFlyweights) {
        for (String state : initialFlyweights) {
            flyweights.put(state, new ConcreteFlyweight(state));
        }
    }
    
    public Flyweight getFlyweight(String key) {
        if (!flyweights.containsKey(key)) {
            flyweights.put(key, new ConcreteFlyweight(key));
        }
        return flyweights.get(key);
    }
    
    public void listFlyweights() {
        System.out.println("Flyweights: " + flyweights.keySet());
    }
}

// Usage
class FlyweightDemo {
    public static void main(String[] args) {
        FlyweightFactory factory = new FlyweightFactory(Arrays.asList("A", "B"));
        
        Flyweight flyweight1 = factory.getFlyweight("A");
        System.out.println(flyweight1.operation("extrinsic1"));
        
        Flyweight flyweight2 = factory.getFlyweight("B");
        System.out.println(flyweight2.operation("extrinsic2"));
        
        Flyweight flyweight3 = factory.getFlyweight("A");
        System.out.println("Same flyweight: " + (flyweight1 == flyweight3));
        
        factory.listFlyweights();
    }
}

File: Estructurales Python/flyweight.py
# Flyweight Pattern
from abc import ABC, abstractmethod

class Flyweight(ABC):
    @abstractmethod
    def operation(self, extrinsic_state):
        pass

class ConcreteFlyweight(Flyweight):
    def __init__(self, intrinsic_state):
        self._intrinsic_state = intrinsic_state
    
    def operation(self, extrinsic_state):
        return f"ConcreteFlyweight: Intrinsic({self._intrinsic_state}), Extrinsic({extrinsic_state})"

class FlyweightFactory:
    _flyweights = {}
    
    def __init__(self, initial_flyweights):
        for state in initial_flyweights:
            self._flyweights[state] = ConcreteFlyweight(state)
    
    def get_flyweight(self, key):
        if key not in self._flyweights:
            self._flyweights[key] = ConcreteFlyweight(key)
        return self._flyweights[key]
    
    def list_flyweights(self):
        return list(self._flyweights.keys())

# Usage
if __name__ == "__main__":
    factory = FlyweightFactory(["A", "B"])
    
    flyweight1 = factory.get_flyweight("A")
    print(flyweight1.operation("extrinsic1"))
    
    flyweight2 = factory.get_flyweight("B")
    print(flyweight2.operation("extrinsic2"))
    
    flyweight3 = factory.get_flyweight("A")
    print(f"Same flyweight: {flyweight1 is flyweight3}")
    
    print(f"Flyweights: {factory.list_flyweights()}")
